Improve tree screen reader accessibility

adomasven opened the pull request:
Addresses all points of #2261

Some notes:
> 4\. When a list is selected, it says "table" instead of "tree view".

I've set the roles for the trees where appropriate to either 'tree' or 'treegrid'. 'treegrid' is announced as table, tree, grid or others depending on screen reading software

> 6\. The item type, colored tags, and non-matching (gray) rows aren't indicated in any way. They should probably be read before or after the title. E.g., "book The Great Gatsby", "non-matching book The Great Gatsby", "book The Great Gatsby tag Class". We'll talk to some people to figure out the best approach here, but getting them into the row would be a good start.

I've added something to address this, but it's not ideal. If you have Item Type column, the item type is read twice, even though i set aria-hidden=true on the Item Type column cell. (coloured) Tags are read at the start and multiple tags are read as "Tag One, Tag Two", etc. 

There doesn't seem to be a good way to make the screen reader make pauses between stuff that is read, although this is probably down to preference and accustomization of screen reader users.

I've tested this on Ubuntu with Orca and it works although it seems to read every selected row twice for some reason. VoiceOver works fairly well, however it does not read level and row information, although maybe that can be changed in the preferences.

dstillman commented:
(Only tested with VoiceOver so far)


1) A row should only be selected when the tree gets focus (e.g., via Tab). This is now selecting a row at all times. Whenever you switch between a collection, the selection should be empty (such that you see the item count in the right-hand pane).

2) After trying this, I think saying the item type after the title is better than saying it before. If you're searching for an item, it's much quicker to hear the title first. Same with colored tags.

adomasven commented:
> 1. A row should only be selected when the tree gets focus (e.g., via Tab). This is now selecting a row at all times. Whenever you switch between a collection, the selection should be empty (such that you see the item count in the right-hand pane).

Ah right. Yeah, this is actually better in many ways in terms of codepaths.

> 2. After trying this, I think saying the item type after the title is better than saying it before. If you're searching for an item, it's much quicker to hear the title first. Same with colored tags.

Probably right. Unfortunately there's no nice way to do this in markup. I've set the item-type icon and tags to have an `aria-label` attribute and there's no simple way to reorder the way they are read, so we might have to override the span with item title that includes this information. It's not bad per-se, but just not neat.

adomasven commented:
Updated. I've also made it announce if an item is retracted, although `retraction.banner` may not be the best string to say there.

dstillman reviewed (commented):
- Is there any way of getting to the column headers via the keyboard? I couldn't figure out any way to do so, even with VoiceOver's separate navigation method. (But I can't do it in the bookmarks manager in the current Firefox either.)
- Now that we're reading the item type as part of the title field, can we disable selection of the item type icon? When I use VoiceOver's navigation to move between elements (theoretically Ctrl-Option-Shift-Down-Arrow to get into the table followed by Ctrl-Option-left/right, though for some reason that doesn't always work), it selects the item type icon but doesn't read anything useful. Same with the twisty (which is operable on the row itself, so doesn't need to be selectable).

adomasven commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -2534,18 +2543,43 @@ var ItemTree = class ItemTree extends LibraryTree {
+		let coloredTags = item.getColoredTags();
+		if (coloredTags.length) {
+			tagSpans = coloredTags.map(x => this._getTagSwatch(x.tag, x.color));
+			tagAriaLabel = Zotero.getString('itemFields.tags') + ' ' + coloredTags.map(x => x.tag).join(',') + '.';;
```
Yeah, I don't like that we need borrow the i18n text from all over the random places. Mostly a naming issue of the i18n strings though.

adomasven commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -823,6 +823,15 @@ var ItemTree = class ItemTree extends LibraryTree {
+			return false;
+		}
+		if (this.selection.count == 0) {
+			this.selection.select(0);
```
I don't remember anymore why it works like this, but I made every behavior of this tree as compatible as possible with the old mozilla tree. We can change this, but it's kind of difficult and perhaps unnecessary. There are a bunch of async things that happen on selection (via notifier and others) and things break in terrible ways if we change selection behavior.

adomasven commented:
> Is there any way of getting to the column headers via the keyboard? I couldn't figure out any way to do so, even with VoiceOver's separate navigation method. (But I can't do it in the bookmarks manager in the current Firefox either.)

I don't know why, but I've never seen the ability to select headers via keyboard shortcuts. Seems like breaking an accessibility pattern (although it does seem dumb since you control a bunch of things about the displayed list from there). Having done some reading around, it's not clear there's a defined pattern for how column hiding and sorting should be implemented, but it seems like a custom setup is in order, like a special button/link that only appears when tabbing for grid sorting/hiding options.

> Now that we're reading the item type as part of the title field, can we disable selection of the item type icon? When I use VoiceOver's navigation to move between elements (theoretically Ctrl-Option-Shift-Down-Arrow to get into the table followed by Ctrl-Option-left/right, though for some reason that doesn't always work), it selects the item type icon but doesn't read anything useful. Same with the twisty (which is operable on the row itself, so doesn't need to be selectable).

Technically when you're navigating the tree with the keyboard the focused element is always the virtualized-table element to implement arrow key navigation properly and all key handlers are attached on that element, so you're never supposed to select individual row elements manually. I will look into a way to disable that completely.

adomasven commented:
> > Now that we're reading the item type as part of the title field, can we disable selection of the item type icon? When I use VoiceOver's navigation to move between elements (theoretically Ctrl-Option-Shift-Down-Arrow to get into the table followed by Ctrl-Option-left/right, though for some reason that doesn't always work), it selects the item type icon but doesn't read anything useful. Same with the twisty (which is operable on the row itself, so doesn't need to be selectable).
> 
> Technically when you're navigating the tree with the keyboard the focused element is always the virtualized-table element to implement arrow key navigation properly and all key handlers are attached on that element, so you're never supposed to select individual row elements manually. I will look into a way to disable that completely.

Yeah, this doesn't seem to work for me at all and it definitely shouldn't. The virtualized-table is disabled for accesibility with `tab-index=-1` which should indicate that it and all elements below it are not accessible. I guess the fact that the role is Table is throwing VoiceOver off.

adomasven commented:
Updated.

Also added a way to open the context menu for items via keyboard. See https://forums.zotero.org/discussion/93436/zotero-beta-html-tree-menu-key-not-working

Sorry it took a while. Somehow I missed the email about your recent comments

dstillman commented:
Can we simulate a more reasonable anchor point for the context menu when triggered via the keyboard? E.g., aligned to the bottom left edge of the row, rather than to the top-left corner of the window?

adomasven commented:
Updated
